### ISSUE TITLE
Possible deadlock condition

### ISSUE THREAD

**Issue by jsonjunky** [dask/distributed#1030]
Hi,
 I keep running into a deadlock in this bit of code in utils.py.

distributed 1.16.1
tornado 4.4.3
dask==0.14.1

`def sync(loop, func, *args, **kwargs):
    """
    Run coroutine in loop running in separate thread.
    """
    if not loop._running:
        try:
            return loop.run_sync(lambda: func(*args, **kwargs))
        except RuntimeError:  # loop already running
            pass

    e = threading.Event()
    main_tid = get_thread_identity()
    result = [None]
    error = [False]

    @gen.coroutine
    def f():
        try:
            if main_tid == get_thread_identity():
                raise RuntimeError("sync() called from thread of running loop")
            yield gen.moment
            result[0] = yield gen.maybe_future(func(*args, **kwargs))
        except Exception as exc:
            logger.exception(exc)
            error[0] = sys.exc_info()
        finally:
            e.set()

    loop.add_callback(f)
    **while not e.is_set():
        e.wait(1000000)**
    if error[0]:
        six.reraise(*error[0])
    else:
        return result[0]`

**Comment by mrocklin** [comment]
Are you able to give more information about what's happening that causes this error?

What you've written here tells me that the `sync` function is waiting on something to complete.  I'm not able to track this down further without knowing more about what it's waiting for.

**Comment by jsonjunky** [comment]
Hi
I don't have much information other than it hangs on that line when I submit work but if I have more workers then it doesn't hang. Does a single worker allow reentrant work requests? Maybe my worker (which creates new task submitted to scheduler) submits more work that gets targeted to it and it hangs? My understanding is workers can submit new tasks within tasks so I wouldn't expect a deadlock scenario to occur.

Outside of that, can you shed some light on "waiting on something to complete"? What things would it be waiting for and why? Then I can look into those conditions. I would expect it to be fully async.

**Comment by mrocklin** [comment]
> Maybe my worker (which creates new task submitted to scheduler)

Information like this is very valuable.  Can you provide a small example that reproduces the kind of error that you're seeing?  

People use Dask in a very broad variety of ways.  It's hard for us to help without more information.

**Comment by jsonjunky** [comment]
I will create a small example and see if I can isolate it. thanks for the quick response. Very nice package you have here and I came from many years frustration with c*lery.

**Comment by mrocklin** [comment]
Thanks for the effort to produce an example.  Hopefully we can provide a
less frustrating experience :)

**Comment by jsonjunky** [comment]
Also. Not sure if this makes a big difference but I'm running scheduler and workers in docker containers using host network. Fyi.

**Comment by mrocklin** [comment]
Shouldn't make a difference as far as I know (assuming that everything is
hooked up correctly).  People certainly use Dask inside Docker containers
in production.

**Comment by jsonjunky** [comment]
I traced this a little more. in utils.py

    @gen.coroutine
    def f():
        try:
            if main_tid == get_thread_identity():
                raise RuntimeError("sync() called from thread of running loop")
            yield gen.moment  <------
            result[0] = yield make_coro()
        except Exception as exc:
            logger.exception(exc)
            error[0] = sys.exc_info()
        finally:
            e.set()

    loop.add_callback(f)


When I create the client, it arrives at this code and enters f() and yields gen.moment.
The finally clause is never called. Is it supposed to be in python 2.7.x?
Unless that e.set() is called, it will enter this infinite loop:

    while not e.is_set():
        e.wait(1000000)

**Comment by jsonjunky** [comment]
I mean, it happens within the first 

client = Client("host:port") 

call. So I can't really say what it's waiting for. I don't know what its doing as I'm only trying to create a client object. It would seem the Tornado IOLoop object is supposed to iterate over the generator produced by 'f' but it is not. I have no idea how Tornado works or what this section of code is supposed to do.

**Comment by jsonjunky** [comment]
I think I got past this problem. I'm fairly sure that when I was trying to print a client.get function then it was hanging on that. Not sure why, but I commented out that print and it flows successfully now.
